docs(adr): [A11y] Offer HTML alternative to our pdf standard documents (#5911)
shopware/platform: adr/2024-12-19-offer-html-alternative-to-our-pdf-standard-document.md

date: 2024-12-19
area: accessibility
tags: [a11y, pdf, html]
---

## Context
To ensure compliance with Web Content Accessibility Guidelines (WCAG), we aim to make Shopware 6's standard document solution A11y-compliant. This will enable continued use of the solution while adhering to legal requirements and making documents accessible to customers with disabilities.

Currently, our PDF generation library, DomPDF, does not meet accessibility standards, which presents a significant challenge for PDF accessibility.

## Decision
In addition to the existing PDF documents generated with DomPDF, we will provide standard document types in an accessible HTML format that fulfills A11y requirements.

### Affected areas
We will integrate HTML document support in the following areas:
1. **Order Detail Page (Admin)**:
    - Preview options:
        - PDF preview
        - HTML preview

2. **Order Detail Page (Storefront)**:
    - Download options:
        - PDF download
        - HTML download

3. **Document Creation**:
    - **Flow Builder**: Automatically generates both PDF and HTML documents.
    - **Bulk Edit**: Automatically generates both PDF and HTML documents.

> 4. **Email Attachments**:

Reviewer: For the last discussion, we'll go with injecting URL to customer access instead of an attachment file, due  many email provider not allow HTML file

Reviewer 2: How can the customer then "download" e.g. his HTML invoice? I guess it's important that these documents can be persisted by the customer. That's why PDF attachments are great, as they are stored inside the email and can be saved to another more persistent location if needed.

A download URL isn't great, because what if that URL isn't working anymore (e.g. Shop domain changed, Shop doesn't exist anymore, ...), but I guess is fine? At least I remember that I received some invoices via download links in my personal life.

Also what is meant by "go with injecting URL to customer access"? Will that URL render HTML in the browser or trigger a download of a `*.html` file?

Reviewer: Hi @MalteJanz , thank your feedback. I already update more details about that in the following thread below.

    - Include the new HTML document format.

5. **E-Invoicing**:
    - Enable document export as HTML.

### Technical implementation
* To support this, we will modify the document entity to store media IDs for both PDF and HTML documents:

```php
$this->addColumn(
    connection: $connection,
    table: 'document',
    column: 'document_media_file_ids',
    type: 'JSON'
);
```

> * In class `Shopware\Core\Checkout\Document\Controller\DocumentController` we will update method `downloadDocument` (`'/api/_action/document/{documentId}/{deepLinkCode}'`) to get a new query parameter `fileType` which can be `pdf` or `html`. If the type is `html` we will return the HTML document.

Reviewer: The route and controller are not very interesting. 

We should add a section on how to protect and secure HTML files (e.g., limit executable content in HTML) and store all those files securely. When we allowed_extensions HTML type

* In class `Shopware\Core\Checkout\Document\SalesChannel\DocumentRoute` we will update method `download` (`'/store-api/document/download/{documentId}/{deepLinkCode}'`) to get a new query parameter `fileType` which can be `pdf` or `html`. If the type is `html` we will return the HTML document.

### Update method render
* Update the render method in all classes extending `AbstractDocumentRenderer` to support rendering HTML documents. For example:

`Shopware\Core\Checkout\Document\Renderer\CreditNoteRenderer`

before:
```php
public function render(array $operations, Context $context, DocumentRendererConfig $rendererConfig): RendererResult
{
    ...
    $html = $this->documentTemplateRenderer->render(
        $template,

>         [

Reviewer: Could we simply pseudo-code here? We just show what is most code impacted: In my opinion, an ADR doesn't need to have detailed coding. 

Ex: 
public function render(array $operations, Context $context, DocumentRendererConfig $rendererConfig): RendererResult
{
    ...
    $html = $this->documentTemplateRenderer->render(..) << that helps me quickly not need to check do we have different constructor parameters or not

---
title: [A11y] Offer HTML alternative to our pdf standard documents
date: 2024-12-19
area: accessibility
tags: [a11y, pdf, html]
---

## Context

To comply with Web Content Accessibility Guidelines (WCAG), we aim to make Shopware 6's document solution accessible (A11y-compliant). This ensures that our solution remains usable while meeting legal requirements and making documents accessible to customers with disabilities. 

Currently, our PDF generation library, DomPDF, does not meet accessibility standards, posing a significant challenge.

## Decision

> We have decided to switch from tagged PDFs to HTML for better accessibility.

Reviewer 2: I wouldn't call it `switch`, as that sounds like we are replacing PDFs. Also we don't have tagged PDFs currently.
```suggestion
We have decided to make HTML documents available in addition to PDF documents, as these are more accessible.
```

- **Better for Accessibility**: HTML is naturally organized, making it easier for accessibility tools to read and present content to people with disabilities.

> - **Lack of Support**: As current PDFs lack support for accessibility. Few tools, especially in PHP, can create accessible tagged PDFs, making it difficult to maintain PDF accessibility.

Reviewer 2: ```suggestion
- **Lack of Support**: As our current PDFs lack support for accessibility. Few tools, especially in PHP, can create accessible tagged PDFs, making it difficult to maintain PDF accessibility.
```

- **Industry Trends**: Many organizations are already moving from PDFs to HTML for accessibility. For example, government websites have been required to meet accessibility standards since the early 2000s. Most of them now use HTML for most of their content because it meets these standards better.

> Switching to HTML aligns with these trends and ensures we are using best practices for accessibility.

Reviewer 2: ```suggestion
Providing HTML documents aligns with these trends and ensures we are using best practices for accessibility.
```

### Affected Areas

We will integrate HTML A11y document support in the following areas:

1. **Document Type Support**:

>     - Support includes `invoice`, `delivery note`, `credit note`, and `cancellation invoice`.

Reviewer 2: ```suggestion
    - Support includes all document types Shopware provides by default, `invoice`, `delivery note`, `credit note`, and `cancellation invoice`. Extensions must adapt themselves.
```

2. **Administration**:
    - **Order Detail Page**: Option to download HTML alongside PDF for each document type.
    - **Document Settings**: Toggle to generate HTML documents.
3. **Storefront**:
    - **Order History**: Customers can access HTML and PDF versions of documents.
4. **Flow Builder**:
    - This setup requires no additional special actions, and merchants can customize file generation for "Generate documents" in the `Document Settings`
5. **Email Delivery**:
    - Enhance the original email by including a link to the HTML document. Customers will need to log in to access the document, and additional guidance will be provided.

>     - As we can’t attach the HTML file directly due to issues with "virus scanners," many email providers do not allow HTML file attachments.

Reviewer 2: ```suggestion
    - We can’t attach the HTML file directly due to issues with "virus scanners", as many email providers do not allow HTML file attachments. Instead we will provide a link inside the Email.
```

            <!-- Headline content -->
        </h1>
    {% endblock %}
    ```

   `src/Core/Framework/Resources/views/documents/style_base_html.css.twig`:
    ```twig
    {% block document_style_html %}
        body {
            max-width: 1200px;
            margin: auto;
            font-size: 14px;
            line-height: 18px;
        }
        ...
    {% endblock %}
    ```

2. **Metadata and Security**:
    - The generation date of the HTML will be "fingerprinted" by adding a metadata header. This allows users to track the creation date of the document.
    - Implement a Content-Security-Policy meta-tag to minimize XSS attack risks, such as disallowing JavaScript and Restricts <base> to the same domain, protecting against base URL manipulation.

   Added new Twig block for metadata`src/Core/Framework/Resources/views/documents/base.html.twig`:
    ```twig
    {% block document_head_meta_protection %}
        <meta http-equiv="Content-Security-Policy" content="script-src 'none'; base-uri 'self';">
        <meta name="date" content="{{ 'now'|date('c') }}">
    {% endblock %}
    ```

#### Core

1. **Abstract Class for Multi-Format Rendering**

>     - We will introduce an abstract class, `src/Core/Checkout/Document/Service/AbstractDocumentTypeRenderer`, to support rendering multiple document types, including `PDF`, `HTML`, and `XML`.

Reviewer 2: Do we already support `XML`? Otherwise I would remove that as an example

>     - We will introduce an abstract class, `src/Core/Checkout/Document/Service/AbstractDocumentTypeRenderer`, to support rendering multiple document types, including `PDF`, `HTML`, and `XML`.

Reviewer: ```suggestion
    - We will introduce an abstract class, `src/Core/Checkout/Document/Service/AbstractDocumentTypeRenderer`, to support rendering multiple document types, including `PDF` and `HTML`.
```

---
title: [A11y] Offer HTML alternative to our pdf standard documents
date: 2024-12-19
area: accessibility
tags: [a11y, pdf, html]
---

## Context
To ensure compliance with Web Content Accessibility Guidelines (WCAG), we aim to make Shopware 6's standard document solution A11y-compliant. This will enable continued use of the solution while adhering to legal requirements and making documents accessible to customers with disabilities.

Currently, our PDF generation library, DomPDF, does not meet accessibility standards, which presents a significant challenge for PDF accessibility.

## Decision

> In addition to the existing PDF documents generated with DomPDF, we will provide standard document types in an accessible HTML format that fulfills A11y requirements.

Reviewer 3: Let's add some thoughts on why we are doing that, in general ADRs should not only state what we are doing, but why we decided to do so

some things I would expect here:
HTML is by design already structured and better suited for a11y than tagged pdfs
Support for generating tagged pdfs is low, currently no PHP lib supports it
SOme institutions are already moving from PDF to HTML for A11y reasons

Maybe @ravenokko can add some thoughts here especially also on the last point

Reviewer 3: We don't necessarily need this in the ADR, but at least when we release it, I think it would be helpful to have a comparison between the formats what are the pros and cons, e.g. regarding a11y, but also: client support, e-invoice support, "trust" etc -> probably more for PM side

Reviewer 4: > SOme institutions are already moving from PDF to HTML for A11y reasons
> Maybe @ravenokko can add some thoughts here especially also on the last point

The most prominent examples are government websites which were required to meet these accessibility requirements ala the BITV 1.0 since the early 2000's. While PDF's haven't completely gone away, you can see most gov websites have shifted the overwhelming majority of their content to HTML.

Reviewer: I have updated the following feedback in the section Decision:

I would resolve that thread. Feel free to open it again if you have any other feedback

### Affected areas
We will integrate HTML document support in the following areas:
1. **Order Detail Page (Admin)**:
    - Preview options:
        - PDF preview
        - HTML preview

2. **Order Detail Page (Storefront)**:
    - Download options:
        - PDF download
        - HTML download

3. **Document Creation**:

>     - **Flow Builder**: Automatically generates both PDF and HTML documents.

Reviewer 3: Will there be an option to say you only want one or the other? 
Whatever the answer, please mention here and if possible also why we do it that way

Reviewer: I have updated the section Flow Builder

This setup requires no additional special actions, and merchants can customize file generation for "Generate documents" in the Document Settings

    - **Bulk Edit**: Automatically generates both PDF and HTML documents.

4. **Email**:
    - Inject a URL to open the HTML document in the browser.

5. **E-Invoicing**:
    - Enable document export as HTML.

### Technical implementation

> * HTML documents will be generated using the same templates as PDF documents.

Reviewer 3: Based on the test to open our current templates generated HTML in the browser it looked quite ugly, probably this means that the existing templates need to be adjusted (not just our default templates, but the custom templates as well right)?
If that assumption is correct, can we:
1. document what needs to change in the templates based on what we needed to for our default templates?
2. built a way to opt-in that a given template is ready to be generated over HTML?

> * HTML documents will be stored as media files, the same way we store PDF files.

Reviewer 3: Is there some way to "fingerprint" the generation date of the HTML into the file? To make sure that the HTML was not generated at runtime (and thus possibly changed after the fact)? As that is the main benefit of using PDFs 

IIRC @ravenokko  mentioned that there are some ways to do that with HTML files during our call 🤔

Reviewer 4: +1 on some sort of timestamp (either in the file, or file metadata) - the tricky thing we need to address is versioning, which isn't an a11y issue so much as a trust/legal issue. PDF's have high trust value because they are viewed as concrete and unchanging, while HTML by design is dynamic. So we do need to clarify how/when these files are generated in addition to being stored, so customers can trust that they won't be changed.

But perhaps we can also emphasize that, since we are still also providing the pdf's, customers can see they will match.

Reviewer 2: +1 at the very least we should have a meta tag in the HTML head with the generation timestamp. As I personally have never received an Invoice as HTML, I would be interested in how other providers deal with this problem, as HTML is a lot easier to modify, in contrast to PDF, and most commonly used for dynamic content and not files on your computer.

For validating that the document wasn't modified at a later point, only this crazy idea that would come to my mind to use some kind of cryptography (checksum / signing) system that validates itself when the document is opened (by executing a little Javascript) and warns the reader that this document was tempered with / isn't signed by the original author anymore. Basically like you can verify that a [JWT token](https://en.wikipedia.org/wiki/JSON_Web_Token) wasn't modified, with some kind of digital signature.

Also as I have written elsewhere, We should make it obvious for the user that this "HTML document" is a file in it's final form which they ideally should download first and then open. Otherwise it's very easy to to think this is just a server generated page and how would they even download that easily?

Reviewer: Thank your feedback,

As suggested, we will add a metadata header to ensure that HTML is not generated at runtime. We will also add some rule Content-Security-Policy meta-tag to reduce the risk of XSS attacks (such as not allowing js, trusted domain)

And during our refinement meeting, it was mentioned that "HTML files don't legally need to be immutable as long as a PDF fulfilling all requirements is provided."

Regarding adding a logic signature check: 
IMHO, it might introduce additional complexity, such as managing keys, storing hashes, and verifying performance with many Document files. Moreover, using JavaScript for this purpose could be risky; it could not be allowed or disabled.

May we simply benefit from HTTPS encryption, which ensures that all data is transmitted securely and the content remains during transmission?

Reviewer 2: > As suggested, we will add a metadata header to ensure that HTML is not generated at runtime.

Thanks, but technically it could still be "generated at runtime" right?😂 But at least the user could download the file and later kind of proof, that it was generated in the past, and they could find out if the document behind the download link got further modified later, e.g. on accident with a bug in our implementation. For that they of course need to inspect the HTML meta tag.

@ravenokko did you found any example HTML invoices from other companies we could take a look at?

> And during our refinement meeting, it was mentioned that "HTML files don't legally need to be immutable as long as a PDF fulfilling all requirements is provided."

That's good to know. Then I don't think we need to care about that and don't have to introduce anything like it.

> May we simply benefit from HTTPS encryption, which ensures that all data is transmitted securely and the content remains during transmission?

Yes, but I think HTTPS should already be a given at this point in time and every shop should have it set up. So nothing really to consider for the implementation, if that was a question?

Reviewer 4: For examples: 

- Payment providers like PayPal/Stripe email links to webpages (if you have a paypal account you can send yourself invoices)
- App Stores email the invoice itself and link to the account
- A lot of the major platforms (Microsoft, Google, Amazon, etc.) will also email a summary with a link to the customer account for things like Azure/Google Cloud/etc.

Reviewer: After discussion, I have updated the following feedback at section Metadata and Security:

I would resolve that thread. Feel free to open it again if you have any other feedback

>   - We will modify the document entity to store media IDs for both PDF and HTML documents:

Reviewer 3: what happens to the existing `documentMediaFileId`? Will we deprecate and remove it? 

As long as the existing Field is there it will always point to the pdf document right? So let's mention that as well

    ```php
    $this->addColumn(
        connection: $connection,
        table: 'document',
        column: 'document_media_file_ids',
        type: 'JSON'
    );
    ```
* Support opening HTML documents in the browser using a deep link code same as PDF documents.

### Update method render

> * Added new generic method `loadHtmlA11y` to support rendering HTML a11y documents in `AbstractDocumentRenderer`:

Reviewer 3: * what does `generic` mean here? Will it be abstract or not? 
* `loadHtmlA11y` -> I would drop the A11y suffix everywhere, `$renderer->loadHtml()` or `$doc->setHtml()` is more readable and clear enough IMHO

> * Update the render method in all classes extending `AbstractDocumentRenderer` to set the html a11y content in the document entity:

Reviewer 3: What happens if e.g. for custom renderer (as our system is generic here) that does not happen right away? I assume that the HTML for the time being is nullable and we handle it gracefully if it is not yet set?

Reviewer 3: And can we also add this point to the `Breaking changes` section with the changes needed to the templates itself?

Reviewer 3: looking at the implementation, i have questions if this is the right place to put the rendering of the HTML, why don't we do it on the level of the `\Shopware\Core\Checkout\Document\Service\DocumentGenerator`, similar to how we access the `PdfRenderer`? 

That way we don't need to adjust the all the abstract renderers 🤔

Reviewer 3: wouldn't it be cleaner to deprecate rendering of the HTML in the abstract renderers:
```
                $html = $this->documentTemplateRenderer->render(
                    $template,
                    [
                        'order' => $order,
                        'config' => $config,
                        'rootDir' => $this->rootDir,
                        'context' => $context,
                    ],
                    $context,
                    $order->getSalesChannelId(),
                    $order->getLanguageId(),
                    $locale->getCode()
                    $locale->getCode(),
                );
 ```
 
 And instead only provide the template options:
 ```
                 $doc->setTemplateOptions([
                    $template,
                    [
                        'order' => $order,
                        'config' => $config,
                        'rootDir' => $this->rootDir,
                        'context' => $context,
                    ],
                    $context,
                    $order->getSalesChannelId(),
                    $order->getLanguageId(),
                    $locale->getCode(),
                ]);
```

Then the pdf renderer is responsible of first generating the HTML for the pdf and the HTML renderer is responsible of directly generating the "finished" HTML?

Reviewer 5: Hi @keulinho, We need two different configs for the renderer template. That's why I need to move the renderer template into the abstract renderers
- with HtmlRenderer, we need to override the config like

<img width="871" alt="Screenshot 2025-01-08 at 19 05 17" src="https://github.com/user-attachments/assets/9d3c70e0-6bdd-458b-988f-f717398971e0" />

`Shopware\Core\Checkout\Document\Renderer\CreditNoteRenderer`
Example:
```php
...
public function render(..): RendererResult
{
    ...
    $doc = new RenderedDocument(...);
    $doc->setHtmlA11y(
        $this->loadHtmlA11y(...),
    );
    ...
}
```

### Security considerations

> * The deep link code to open HTML documents will be used to authenticate the user and ensure that the user has access to the document.

Reviewer 2: I guess you could reuse the same logic as for downloading PDF documents? But I have another thing in my mind that is missing here:

As you are dealing with HTML, which potentially can contain malicious code, the worst case scenario would be in my opinion:
- user clicks link to view HTML document
- browser opens and renders html under the shops domain
- JS code executes under the shops domain name
- We basically enabled [cross site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting)

That's why I would strongly recommend only dealing with files (`*.pdf` or `*.html`) in a way that only allows to download them. The user can then open the locally stored `*.html` in a browser to view it.

Reviewer: Hi @MalteJanz 

Thank your feedback. After discussing with Juri and Tim, we have an idea to go in the same direction as you suggested:

For email, we will send a new email along with the original email with the content clearly explained and in the directions:

FIY:

```
"Dear {customer.first_name}{customer.last_name},

A new document has been added to your order {OrderNumber}. You should have received the original document in a separate mail.

For better accessibility, we also provide an HTML version of this document here: {Link to HTML file}

For data protection reasons, the HTML version requires a login.

In case of a guest order, you can use your mail address and postal code of the billing address."
```

As you said, when the user clicks on the link, it will lead to the shop. From here, the user can view or download the content.

Reviewer 3: Is there a reason why we not add the HTML as a second attachment in the original mail?

I would expect if a document is also generated as HTML to have an option to add that as well as attachment in this dialogue: 
![Screenshot 2025-01-07 at 14 01 37](https://github.com/user-attachments/assets/f8171732-58df-4e83-92d5-6061fb4bb1a3)

Reviewer 3: We can additionally provide the link to the HTML inside the email template variables, so it can be used there as well, but not sure if we should add another email flow by default for this

Reviewer: Hi @keulinho, many email systems block HTML file attachments because of virus scanners, so we can't attach them directly.

The first idea was to add a link to the HTML file in existing emails, like the Order Confirmation. However:

1. We can’t be sure, in all cases, that the documents actually already exist when the mail (e.g. Order Confirmation) has been sent. 
2. We also considered injecting this content using a migration default email template, which could be more complicated and confuse customers about the default templates. 

Therefore, Juri and Tim suggest we probably need a new mail template that contains the HTML document link. Customers can access the HTML version of the documents through this link. The email should clearly explain the purpose and provide an understandable reason for this approach.

Additionally, we can provide an option setting for this feature in the Document Settings.

Does this work for you, too?

Reviewer 2: I would also vote against introducing a completely new email + template just for sending the a11y link to the customer, as mentioned here: https://github.com/shopware/shopware/pull/5834#discussion_r1907324177

If we can't override existing templates, because they might got adjusted by the merchant already, we should think about other solutions to make them properly upgrade with the shopware major (see linked comment above for one idea).

In the end this decision and the reasoning behind it should also be documented in this ADR

Reviewer: After discussion, I have updated the following feedback and section email Delivery and Email Migration.

I would resolved that thread. Feel free open it again if you have any other feedback
